pyrealm/demography/t_model_functions.py

```diff
@@ -78,6 +78,56 @@ def calculate_heights(
     return h_max * (1 - np.exp(-a_hd * dbh / h_max))
 
 
+def calculate_dbh_from_height(
+    h_max: NDArray[np.float32],
+    a_hd: NDArray[np.float32],
+    stem_height: NDArray[np.float32],
+    validate: bool = True,
+) -> NDArray[np.float32]:
+    r"""Calculate diameter at breast height from stem height under the T Model.
+
+    This function inverts the normal calculation of stem height (:math:`H`) from
+    diameter at breast height (DBH, :math:`D`) in the T Model (see
+    :meth:`~pyrealm.demography.t_model_functions.calculate_heights`). This is a helper
+    function to allow users to convert known stem heights for a plant functional type,
+    with maximum height (:math:`H_{m}`) and initial slope of the height/diameter
+    relationship (:math:`a`) into the expected DBH values.
+
+    .. math::
+
+         D = \frac{H \left( \log \left(\frac{H}{H_{m}-H}\right)\right)}{a}
+
+    Warning:
+        Where the stem height is greater than the maximum height for a PFT, then
+        DBH is undefined and the return array will contain `np.nan`. Where the
+        stem height equals the maximum height, the model predicts an infinite stem
+        diameter: the `h_max` parameter is the asymptotic maximum stem height of an
+        exponential function. Similarly, heights very close to the maximum height may
+        lead to unrealistically large predictions of DBH.
+
+    Args:
+        h_max: Maximum height of the PFT
+        a_hd: Initial slope of the height/diameter relationship of the PFT
+        stem_height: Stem height of individuals
+        validate: Boolean flag to suppress argument validation
+    """
+
+    if validate:
+        _validate_t_model_args(pft_args=[h_max, a_hd], size_args=[stem_height])
+
+    # The equation here blows up in a couple of ways:
+    # - H > h_max leads to negative logs which generates np.nan with an invalid value
+    #   warning. The np.nan here is what we want to happen, so the warning needs
+    #   suppressing.
+    # - H = h_max generates a divide by zero which returns inf with a warning. Here the
```

Comment:
I'm not quite sure why its desirable for the np.nan to be generated.

Would it make sense to add another optional argument to allow the user to an alternative behaviour where the code saturates any stem height at the max height for that PFT?

There could be three alternative behaviours desired the function which could be chosen by the user:
1) Allow np.nan to be calculated and suppress errors (current implementation)
2) Allow np.nan to be calculated and raise errors
3) Force saturation of stem height instead of np.nan

This is well outside my field of expertise so feel free to push back :)

Reply:
Aside from this query, code looks fine. Happy to approve

Reply:
In the existing function for the  DBH to H calculation is fine, $H \to H_{max}$ as $D \to \inf$. This new function inverts that calculation from H to DBH, and so when $H >= H_{max}$, $D$ is genuinely infinite or undefined and so there isn't a sensible asymptotic $D$ to saturate to. I did initially think there was some thing we could do, but we do genuinely need to set `np.nan` for impossible trees.